core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSink.java

```diff
@@ -46,11 +46,15 @@
 
     /**
      * Notifies the exchange sink that no more data will be appended
+     *
+     * @return future that will be resolved when the finish operation either succeeds or fails
      */
-    void finish();
+    CompletableFuture<?> finish();
```

Comment:
what should `finish` do when `abort` is already running. Contract is undefined here

Reply:
> what should finish do when abort is already running? 

`finish` should never be called after abort. If it is - it's a bug. Let me document it. 

> Contract is undefined here what should finish do when another finish is running?

`finish` shouldn't be called when another `finish` is running. If it is - it's a bug. I will also document that.

Reply:
> finish should never be called after abort. If it is - it's a bug. Let me document it.

I think it can in current code. Scenario
1. thread1: `setNoMorePages` is called, but `stateMachine.noMorePages()` is not yet set
2. thread2: `destroy` is called and `abort` is called
3. thread1: resumes execution and `finish` is called (after `abort`)